File: packages/typescript/src/catalogs/sepomex/codigos-postales-completo.ts
/**
 * SEPOMEX - Complete Mexican Postal Codes Catalog
 * All ~150,000 postal codes
 *
 * WARNING: This catalog is very large and should be used with caution.
 * Consider using pagination or filters when displaying results.
 */

import { loadCatalog } from '../../utils/catalog-loader';
import type { PostalCode } from '../../types';

export class CodigosPostalesCompleto {
  private static _data: PostalCode[] | null = null;

  /**
   * Load complete postal codes data (lazy loading with caching)
   * WARNING: This loads ~150,000 postal codes into memory (~15-20 MB)
   */
  private static getData(): PostalCode[] {
    if (!this._data) {
      this._data = loadCatalog<PostalCode>('sepomex/codigos_postales_completo.json');
    }
    return this._data;
  }

  /**
   * Get all postal codes (~150,000 records)
   * WARNING: This returns a very large array
   */
  static getAll(): PostalCode[] {
    return this.getData();
  }

  /**
   * Get postal code information
   */
  static getByCp(cp: string): PostalCode[] {
    return this.getData().filter(postal => postal.cp === cp || postal.codigo_postal === cp);
  }

  /**
   * Get state name from postal code
   */
  static getEstado(cp: string): string | undefined {
    const postal = this.getData().find(p => p.cp === cp || p.codigo_postal === cp);
    return postal?.estado;
  }

  /**
   * Get municipality from postal code
   */
  static getMunicipio(cp: string): string | undefined {
    const postal = this.getData().find(p => p.cp === cp || p.codigo_postal === cp);
    return postal?.municipio;
  }

  /**
   * Search by municipality (returns paginated results)
   */
  static getByMunicipio(municipio: string, limit: number = 100): PostalCode[] {
    const searchTerm = municipio.toUpperCase();
    const results = this.getData().filter(postal =>
      postal.municipio.toUpperCase().includes(searchTerm)
    );
    return results.slice(0, limit);
  }

  /**
   * Search by state (returns paginated results)
   */
  static getByEstado(estado: string, limit: number = 1000): PostalCode[] {
    const searchTerm = estado.toUpperCase();
    const results = this.getData().filter(postal =>
      postal.estado.toUpperCase() === searchTerm
    );
    return results.slice(0, limit);
  }

  /**
   * Search by settlement (asentamiento) with pagination
   */
  static searchByAsentamiento(asentamiento: string, limit: number = 100): PostalCode[] {
    const searchTerm = asentamiento.toUpperCase();
    const results = this.getData().filter(postal =>
      postal.asentamiento.toUpperCase().includes(searchTerm)
    );
    return results.slice(0, limit);
  }

  /**
   * Validate postal code exists
   */
  static isValid(cp: string): boolean {
    return this.getData().some(postal => postal.cp === cp || postal.codigo_postal === cp);
  }

  /**
   * Get total count of postal codes
   */
  static getTotalCount(): number {
    return this.getData().length;
  }

  /**
   * Get all unique postal codes
   */
  static getUniqueCPs(): string[] {
    const cpsSet = new Set(this.getData().map(p => p.cp || p.codigo_postal!));
    return Array.from(cpsSet).sort();
  }

  /**
   * Get count of postal codes by state
   */
  static getCountByEstado(estado: string): number {
    const searchTerm = estado.toUpperCase();
    return this.getData().filter(postal =>
      postal.estado.toUpperCase() === searchTerm
    ).length;
  }

  /**
   * Get all settlements (asentamientos) for a postal code
   */
  static getAsentamientos(cp: string): string[] {
    return this.getData()
      .filter(postal => postal.codigo_postal === cp)
      .map(postal => postal.asentamiento);
  }

  /**
   * Search with multiple criteria (with pagination)
   */
  static search(criteria: {
    cp?: string;
    estado?: string;
    municipio?: string;
    asentamiento?: string;
    limit?: number;
  }): PostalCode[] {
    const limit = criteria.limit ?? 100;
    let results = this.getData();

    if (criteria.cp) {
      results = results.filter(p => p.codigo_postal === criteria.cp);
    }
    if (criteria.estado) {
      const estado = criteria.estado.toUpperCase();
      results = results.filter(p => p.estado.toUpperCase() === estado);
    }
    if (criteria.municipio) {
      const municipio = criteria.municipio.toUpperCase();
      results = results.filter(p => p.municipio.toUpperCase().includes(municipio));
    }
    if (criteria.asentamiento) {
      const asentamiento = criteria.asentamiento.toUpperCase();
      results = results.filter(p => p.asentamiento.toUpperCase().includes(asentamiento));
    }

    return results.slice(0, limit);
  }

  /**
   * Get statistics about the catalog
   */
  static getStatistics(): {
    totalPostalCodes: number;
    uniquePostalCodes: number;
    states: number;
    municipalities: number;
  } {
    const data = this.getData();
    const uniqueCPs = new Set(data.map(p => p.codigo_postal));
    const uniqueStates = new Set(data.map(p => p.estado));
    const uniqueMunicipalities = new Set(data.map(p => `${p.estado}:${p.municipio}`));

    return {
      totalPostalCodes: data.length,
      uniquePostalCodes: uniqueCPs.size,
      states: uniqueStates.size,
      municipalities: uniqueMunicipalities.size
    };
  }
}

File: packages/dart/lib/src/catalogs/sepomex/codigos_postales_completo.dart
/// SEPOMEX Códigos Postales Completo Catalog
///
/// Provides access to Mexican postal codes with enhanced address data.
/// Includes settlement type, zone (urban/rural), and state/municipality codes.
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';
import 'package:catalogmx/src/utils/text_utils.dart';

/// SEPOMEX Postal Codes Completo Catalog (enhanced data)
class SepomexCodigosPostalesCompleto {
  static List<Map<String, dynamic>>? _data;
  static Map<String, List<Map<String, dynamic>>>? _byCP;
  static Map<String, List<Map<String, dynamic>>>? _byState;
  static Map<String, List<Map<String, dynamic>>>? _byMunicipio;
  static Map<String, List<Map<String, dynamic>>>? _byZona;

  /// Loads the postal codes completo data (lazy loading due to large size)
  static void _loadData() {
    if (_data != null) return;

    _data =
        BaseCatalog.loadJsonDataSync('sepomex/codigos_postales_completo.json');

    // Build lookup maps
    _byCP = {};
    _byState = {};
    _byMunicipio = {};
    _byZona = {};

    for (final code in _data!) {
      final cp = code['cp'] as String?;
      final estado = code['estado'] as String?;
      final municipio = code['municipio'] as String?;
      final zona = code['zona'] as String?;

      if (cp != null) {
        _byCP![cp] ??= [];
        _byCP![cp]!.add(code);
      }

      if (estado != null) {
        _byState![estado] ??= [];
        _byState![estado]!.add(code);
      }

      if (municipio != null) {
        _byMunicipio![municipio] ??= [];
        _byMunicipio![municipio]!.add(code);
      }

      if (zona != null) {
        _byZona![zona] ??= [];
        _byZona![zona]!.add(code);
      }
    }
  }

  /// Gets all postal codes (Warning: Large dataset)
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Gets postal code information by CP (may return multiple settlements)
  static List<Map<String, dynamic>> getByCP(String codigoPostal) {
    _loadData();
    return List.from(_byCP![codigoPostal] ?? []);
  }

  /// Gets all postal codes for a state (accent-insensitive)
  static List<Map<String, dynamic>> getByState(String estado) {
    _loadData();
    // Try exact match first
    final exactMatch = _byState![estado];
    if (exactMatch != null && exactMatch.isNotEmpty) {
      return List.from(exactMatch);
    }

    // Fallback to normalized search
    final normalized = normalizeText(estado);
    for (final entry in _byState!.entries) {
      if (normalizeText(entry.key) == normalized) {
        return List.from(entry.value);
      }
    }
    return [];
  }

  /// Gets all postal codes for a municipality
  static List<Map<String, dynamic>> getByMunicipio(String municipio) {
    _loadData();
    // Try exact match first
    final exactMatch = _byMunicipio![municipio];
    if (exactMatch != null && exactMatch.isNotEmpty) {
      return List.from(exactMatch);
    }

    // Fallback to normalized search
    final normalized = normalizeText(municipio);
    for (final entry in _byMunicipio!.entries) {
      if (normalizeText(entry.key) == normalized) {
        return List.from(entry.value);
      }
    }
    return [];
  }

  /// Gets postal codes by zone (Urbano or Rural)
  static List<Map<String, dynamic>> getByZona(String zona) {
    _loadData();
    return List.from(_byZona![zona] ?? []);
  }

  /// Searches postal codes by settlement name (accent-insensitive)
  static List<Map<String, dynamic>> searchByAsentamiento(
    String query, {
    String? codigoPostal,
    String? estado,
  }) {
    _loadData();
    final normalized = normalizeText(query);

    var searchList = _data!;
    if (codigoPostal != null) {
      searchList = getByCP(codigoPostal);
    } else if (estado != null) {
      searchList = getByState(estado);
    }

    return searchList.where((code) {
      final asentamiento = code['asentamiento'] as String? ?? '';
      return normalizeText(asentamiento).contains(normalized);
    }).toList();
  }

  /// Gets postal codes by settlement type (Colonia, Fraccionamiento, etc.)
  static List<Map<String, dynamic>> getByTipoAsentamiento(String tipo) {
    _loadData();
    final normalized = normalizeText(tipo);
    return _data!.where((code) {
      final tipoAsentamiento = code['tipo_asentamiento'] as String? ?? '';
      return normalizeText(tipoAsentamiento).contains(normalized);
    }).toList();
  }

  /// Validates if a postal code exists
  static bool isValid(String codigoPostal) {
    _loadData();
    return _byCP!.containsKey(codigoPostal);
  }

  /// Gets total count of postal codes
  static int get count {
    _loadData();
    return _data!.length;
  }

  /// Clears cached data to free memory
  static void clearCache() {
    _data = null;
    _byCP = null;
    _byState = null;
    _byMunicipio = null;
    _byZona = null;
    BaseCatalog.clearCacheFor('sepomex/codigos_postales_completo.json');
  }
}
